### ISSUE TITLE
stack_cap failure with Qemu-CHERI

### ISSUE THREAD

**Issue by staceyson** [CTSRD-CHERI/cheri-c-tests#2]
stack_cap causes a cap length violation in the start up of the test.  I am wondering if this is a problem with Qemu-CHERI or the test code.  All the other tests pass.

The register dump is below.  The length violation happens in very early in test_setup():

```
0000000120003428 <test_setup>:
   120003428:   67bdff40        daddiu  sp,sp,-192
   12000342c:   67a400b8        daddiu  a0,sp,184
   120003430:   fa4be880        csc     c18,sp,128(c11)  <— length violation here
   120003434:   fa2be860        csc     c17,sp,96(c11)
   120003438:   ebcb2003        csd     s8,a0,0(c11)
```

Looking at the instruction trace, it seems the $sp might be incorrect:

```
0x0000000120095194:  nop
0x0000000120003428:  daddiu     sp,sp,-192
    Write sp = fffffffffff1ef80
0x000000012000342c:  daddiu     a0,sp,184
    Write a0 = fffffffffff1f038
0x0000000120003430:  csc        c18,sp,465(c11)
```

```
root@:~/src/cheri-c-tests # ./stack_cap 
CHERI cause: ExcCode: 0x01 RegNum: $c11 (length violation)
$c00: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:0 t:0
$c01: v:1 s:0 p:7fff8017 b:0000000000000000 l:0000010000000000 o:120095188 t:0
$c02: v:1 s:0 p:7fff807d b:0000007fffffe2bc l:0000000000000004 o:0 t:0
$c03: v:1 s:0 p:7fff807d b:0000007fffffe400 l:0000000000000040 o:0 t:0
$c04: v:1 s:0 p:7fff807d b:0000007fffffe440 l:0000000000000280 o:0 t:0
$c05: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c06: v:1 s:0 p:7fff807d b:0000007fffffe058 l:0000000000000008 o:0 t:0
$c07: v:1 s:0 p:7fff807d b:0000007fffffe050 l:0000000000000008 o:0 t:0
$c08: v:1 s:0 p:7fff807d b:0000007fffffe260 l:0000000000000020 o:0 t:0
$c09: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c10: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c11: v:1 s:0 p:7fff807d b:0000007ffff7f000 l:000000000007f360 o:0 t:0
$c12: v:1 s:0 p:7fff8017 b:0000000000000000 l:0000010000000000 o:120003428 t:0
$c13: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c14: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c15: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c16: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c17: v:1 s:0 p:7fff8017 b:0000000000000000 l:0000010000000000 o:120095198 t:0
$c18: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c19: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c20: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c21: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c22: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c23: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c24: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c26: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:0 t:0
$c31: v:1 s:0 p:7fff8017 b:0000000000000000 l:0000010000000000 o:120003430 t:0
Jan 23 16:56:59  kernel: USER_CHERI_EXCEPTION: pid 985 tid 100053 (stack_cap), uid 0: CP2 fault (type 0x32)
Jan 23 16:56:59  kernel: Trapframe Register Dump:
Jan 23 16:56:59  kernel: zero: 0        at: 0x120003428 v0: 0x1200c9ab0 v1: 0x1
Jan 23 16:56:59  kernel: a0: 0xfffffffffff1f038 a1: 0x7f238     a2: 0x7f048     a3: 0x7f048
Jan 23 16:56:59  kernel: a4: 0  a5: 0   a6: 0   a6: 0
Jan 23 16:56:59  kernel: t0: 0  t1: 0   t2: 0   t3: 0
Jan 23 16:57:00  kernel: t8: 0  t9: 0x1200950e0 s0: 0x1200c9ab0 s1: 0
Jan 23 16:57:00  kernel: s2: 0  s3: 0   s4: 0   s5: 0
Jan 23 16:57:00  kernel: s6: 0  s7: 0   k0: 0   k1: 0
Jan 23 16:57:00  kernel: gp: 0  sp: 0xfffffffffff1ef80  s8: 0xfffffffffff1f040  ra: 0
Jan 23 16:57:00  kernel: sr: 0x408084b3 mullo: 0        mulhi: 0        badvaddr: 0x120003430
Jan 23 16:57:00  kernel: cause: 0x48    pc: 0x120003430
Signal 34 (core dumped)
```

**Comment by davidchisnall** [comment]
Your `$sp` value does indeed look suspicious.  This should be an offset into `$c11`, but `0xfffffffffff1ef80` looks very negative.  This is odd if the other tests pass because:
- `test_setup` is called directly from `main` and if  the stack is not set up on entry to main, this would probably cause problems in all of the tests.
- `test_setup` is called from `main` in _all_ of the tests.
- This `$sp` value should have caused a length violation in the return address spill code in `main`, so you shouldn't have been able to get to this point if it had been wrong on entry to `main`.

Do you have a trace of this running that I can look at?

**Comment by staceyson** [comment]
I uploaded two instruction traces to https://people.freebsd.org/~sson/mips/traces/.  stack_cap.trace.xz is cheritrace compatible.  There are some bugs with Qemu-cheri's cheritrace generation still.  For example, the ASID is always zero.  stack_cap.log.xz is a compressed text file of the instruction trace in Qemu-Cheri's original format.

**Comment by davidchisnall** [comment]
Can you upload the test binary too please?

**Comment by staceyson** [comment]
Uploaded to  https://people.freebsd.org/~sson/mips/traces/ as well.

It seems that the $sp goes negative at the beginning of main():

```
0x00000001200950e0:  daddiu     at,zero,1  <--- main()
    Write at = 0000000000000001
0x00000001200950e4:  dsll32     at,at,0xf
    Write at = 0000800000000000
0x00000001200950e8:  daddiu     at,at,-11
    Write at = 00007ffffffffff5
0x00000001200950ec:  dsll       at,at,0x11
    Write at = ffffffffffea0000
0x00000001200950f0:  daddiu     at,at,-224
    Write at = ffffffffffe9ff20
0x00000001200950f4:  daddu      sp,sp,at
    Write sp = fffffffffff1f040
```

**Comment by davidchisnall** [comment]
It looks as if the bug is that `$c11` is initialised with too small a length:

```
c11 v:1 u:0 perms:0x0000807d type:0x80000000 offset:0x0000000000000000 base:0x0000007ffff7f000 length:0x000000000007f360    
```

The test is trying to allocate over 1MB of stack space, but the stack is only 256KB.

**Comment by brooksdavis** [comment]
Fixed in CTSRD-CHERI/cheribsd@580cfb8b39a685ad586e0f870213031042842554.
